DigitalLearningSolutions.Data.Tests/Helpers/LearningHubSsoSecurityHelperTests.cs

```diff
@@ -0,0 +1,185 @@
+
+        [Test]
+        public void GenerateHash_returns_hashes_for_timestamps_accurate_to_the_second()
+        {
+            // Given
+            var now = DateTime.UtcNow.Date;
+            var stateString = "stateString";
+            var clockService = new BinaryClockService(now, now.AddMilliseconds(999));
+            var helper = new LearningHubSsoSecurityHelper(clockService, Config);
+
+            // When
+            var hash1 = helper.GenerateHash(stateString);
+            var hash2 = helper.GenerateHash(stateString);
+
+            // Then
+            hash1.Should().Be(hash2);
+        }
+
+        [Test]
+        public void GenerateHash_returns_different_hashes_for_different_state_input_strings()
+        {
+            // Given
+            var now = DateTime.UtcNow;
+            var stateString = "stateString";
+            var differentStateString = "stateStrinh";
+            var clockService = new BinaryClockService(now, now.AddSeconds(1));
+            var helper = new LearningHubSsoSecurityHelper(clockService, Config);
+
+            // When
+            var hash1 = helper.GenerateHash(stateString);
+            var hash2 = helper.GenerateHash(differentStateString);
+
+            // Then
+            hash1.Should().NotBe(hash2);
+        }
+
+        [Test]
+        public void GenerateHash_returns_different_hashes_for_different_secret_keys()
+        {
+            // Given
+            var now = DateTime.UtcNow;
+            var stateString = "stateString";
+            var clockService = new BinaryClockService(now, now);
+            var helper = new LearningHubSsoSecurityHelper(clockService, Config);
+
+            var alternateConfig = A.Fake<IConfiguration>();
+
+            A.CallTo(() => alternateConfig["LearningHubSSO:ToleranceInSeconds"]).Returns(TestTolerance.ToString());
+            A.CallTo(() => alternateConfig["LearningHubSSO:HashIterations"]).Returns(TestIterations.ToString());
+            A.CallTo(() => alternateConfig["LearningHubSSO:ByteLength"]).Returns(TestLength.ToString());
+            A.CallTo(() => alternateConfig["LearningHubSSO:SecretKey"]).Returns("open sesame");
+
+            var helper2 = new LearningHubSsoSecurityHelper(clockService, alternateConfig);
+
+            // When
+            var hash1 = helper.GenerateHash(stateString);
+            var hash2 = helper2.GenerateHash(stateString);
+
+            // Then
+            hash1.Should().NotBe(hash2);
+        }
+
+        [Test]
+        public void VerifyHash_returns_true_for_hashed_created_within_tolerance_time([Range(-TestTolerance, TestTolerance, 1)] int delay)
+        {
+            // Given
+            var now = DateTime.UtcNow;
+            var stateString = "stateString";
+            var clockService = new BinaryClockService(now, now.AddSeconds(delay));
+            var helper = new LearningHubSsoSecurityHelper(clockService, Config);
+
+            // When
+            var hash1 = helper.GenerateHash(stateString);

```

Comment:
Nitpick that in this test and the one below there's only one hash, so you can just call it `hash` instead of `hash1`.